Improve documentation for Antrea-native policy (#3512)
antrea-io/antrea: docs/antrea-network-policy.md

> Antrea release 0.9.x introduced support for 5 static tiers. These static tiers

Reviewer: Is there a specific reason for using "tiers" instead of "Tiers" here and in the section title? Looks like all the others are replaced.

Author: This intricate capitalization scheme is to indicate that before 0.9.x "tier" is merely a field or a concept for the Antrea-native policies, while since 1.0 it became a K8s resource, hence capitalized.

Reviewer: Understood👌

Any Antrea-native policy CRD referencing a static tier in its spec will now internally
reference the corresponding Tier resource, thus maintaining the order of enforcement.

> The default Tiers are created as follows in the relative order of

Reviewer 2: should we use "static" or "default" consistently?

Author: Done

    2021/06/24 23:56:41.346165 AntreaPolicyEgressRule AntreaNetworkPolicy:default/test-anp Drop 44900 10.10.1.65 35402 10.0.0.5 80 TCP 60 [3 packets in 1.011379442s]
```

> Fluentd can be used to assist with analyzing the logs. Refer to the

Reviewer 2: s/analyzing/collecting and analyzing/

### Behavior of *to* and *from* selectors

The following selectors can be specified in an ingress `from` section or egress `to`

> section for selecting networking peers of the policy:

Reviewer 2: ```suggestion
section when defining networking peers for policy rules.
```
What do you think?

Author: Changed

are grouped as `ingress` "sources" or `egress` "destinations". Cannot be set
with `namespaces` field.

**podSelector** and **namespaceSelector**: A single to/from entry that
specifies both namespaceSelector and podSelector selects particular Pods within
particular Namespaces.

**namespaces**: The `namespaces` field allows users to perform advanced matching on
Namespace objects which cannot be done via label selectors. Refer to 
[this section](#matching-self-namespace-for-appliedtos) for more details, and
[this sample yaml](#acnp-for-strict-namespace-isolation) for usage.

**group**: A `group` refers to a ClusterGroup to which this ingress/egress peer, or

> an `appliedTo` must resolve to. More information on ClusterGroups can be found [here](#clustergroup).

Reviewer 2: nit: maybe replace `here` by `this section` for consistency in the links :)

**serviceAccount**: This selects all Pods that have been assigned by the ServiceAccount

> specified. For more information on its usage, refer to [this section](#serviceaccount-based-selection).

Reviewer 2: > This selects all Pods that have been assigned by the ServiceAccount specified

Doesn't sound correct. I suggest:

> This selects all the Pods which have been assigned a specific ServiceAccount

Author: Done

### Ordering based on Tier priority

> With the introduction of Tiers, Antrea Policies, like ClusterNetworkPolicies,

Author: done

policy rules are realized by OpenFlow, and how the priority of flows reflects the
order in which they are enforced.

> ## Advanced peer selection mechanisms of Antrea-native Policy

Reviewer 2: ```suggestion
## Advanced peer selection mechanisms of Antrea-native Policies
```

        name: AllowToCoreDNS
```

The above example allows all Pods from Namespace "default" to connect to all "kube-dns"
Pods from Namespace "kube-system" on TCP port 53.

> ### Matching self Namespace for appliedTos

Reviewer 2: section header is hard to understand IMO

maybe "Selecting Pods in the Namespace in which a policy rule is applied with `Self`"

Author: I would like to keep it a little more concise. How does "Selecting Pods in the same Namespace with `Self`" sound?

Reviewer: Not related to this PR, but if we extend to `NotSelf`, shall we open a new section "### Selecting ... with `NotSelf`", or attach with this section since it introduces namespaces matching?

Author: We could rename the section to `Advanced Namespace matching` and list two sub-sections for `Self` and `NotSelf`?

The `namespaces` field allows users to perform advanced matching on Namespace objects that cannot 
be done via label selectors. Currently, the `namespaces` field has only one matching strategy, `Self`. 
If set to `Self`, for each Pod affected by `appliedTo` of the policy/rule, this field selects the same
Namespace that the Pod is in. It enables policy writers to create per-Namespace rules within a

> single policy.

Reviewer 2: I feel like the previous version of this paragraph had a bit more details?

Author: Yes there was a more verbose sentence, which I find hard to understand when I read it now (even though I wrote it myself). That's why I put two addition sample specs, one with just the `namespaces` field and the other being the `namespaces`+`podSelector` to illustrate how it works. I was hoping that it will be more readable.

Consider a minimalistic cluster, where there are only three Namespaces labeled ns=x, ns=y and ns=z.
Inside each of these Namespaces, there are three Pods labeled app=a, app=b and app=c.

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: ClusterNetworkPolicy
metadata:
  name: allow-self-ns
spec:
  priority: 1
  tier: platform
  appliedTo:
    - namespaceSelector: {}
  ingress:
    - action: Allow
      from:
        - namespaces:
            match: Self
    - action: Deny
  egress:
    - action: Allow
      to:
        - namespaces:
            match: Self
    - action: Deny
```

The policy above ensures that x/a, x/b and x/c can communicate with each other, but nothing else

> (unless there are higher precedenced policies which says otherwise). Same for Namespaces y and z.

Reviewer 2: s/says/say

```yaml
apiVersion: crd.antrea.io/v1alpha1
kind: ClusterNetworkPolicy
metadata:
  name: deny-self-ns-a-to-b
spec:
  priority: 1
  tier: securityops
  appliedTo:
    - namespaceSelector: {}
      podSelector:
        matchLabels:
          app: b
  ingress:
    - action: Deny
      from:
        - namespaces:
            match: Self
          podSelector:
            matchLabels:
              app: a
```

The `deny-self-ns-a-to-b` policy ensures that traffic from x/a -> x/b, y/a -> y/b and z/a -> z/b are

> denied. It can be used in conjunction with the `allow-self-ns` policy. If both policies are applied,

Reviewer 2: either "traffic from x/a to x/b, y/a to y/b and z/a to z/b is denied"
or
"traffic x/a -> x/b, y/a -> y/b and z/a -> z/b is denied"

Author: Done

the only other Pod that x/a can reach in the cluster will be Pod x/c.

These two policies shown above are for demonstration purposes only. For more realistic usage of the
`namespaces` field, refer to this [sample](#acnp-for-strict-namespace-isolation) YAML in the previous

> section. This field is optional and cannot be set along with a `namespaceSelector` within the same peer.

Reviewer 2: maybe this last sentence should come before the examples?

Author: Done

        name: AllowToCoreDNS
```

The above example allows all Pods from Namespace "default" to connect to all "kube-dns"
Pods from Namespace "kube-system" on TCP port 53.

### Selecting Pods in the same Namespace with Self

The `namespaces` field allows users to perform advanced matching on Namespace objects that cannot
be done via label selectors. Currently, the `namespaces` field has only one matching strategy, `Self`.

> If set to `Self`, for each Pod affected by `appliedTo` of the policy/rule, this field selects the same

Reviewer 2: I agree that the previous version had issues, but maybe this is not clear enough. What about:

> If set to `Self`, for each Pod targeted by the `appliedTo` of the policy/rule, this field will cause the rule to select endpoints in the same Namespace as that Pod.

Author: Done

rule to select endpoints in the same Namespace as that Pod. It enables policy writers to create per-Namespace
rules within a single policy.

> This field is optional and cannot be set along with a `namespaceSelector` within the same peer.

Author: fixed

      matchLabels:
        app: client
  egress:
  - action: Allow
    to:
      - fqdn: "*foobar.com"
    ports:
      - protocol: TCP
        port: 8080
  - action: Drop      # Drop all other egress traffic, in-cluster or out-of-cluster 
```

The above example allows all traffic destined to any FQDN that matches the wildcard expression

> `*foobar.com` on port 8080, originating from any Pod with label `app` set to `client`

Reviewer 2: I notice that we don't write anywhere that `*` can match multiple subdomains (i.e. it can match dots in the domain name). For example, IIRC `*foobar.com` will match `x.foobar.com`, `x.x.foobar.com`, etc.
We should probably also say that `*` can match 0 or more characters, so `*foobar.com` will also match `foobar.com`

These statements are based on my understanding, so to be verified.

Author: Added clarification

      matchLabels:
        app: client
  egress:
  - action: Allow
    to:
      - fqdn: "*foobar.com"
    ports:
      - protocol: TCP
        port: 8080
  - action: Drop      # Drop all other egress traffic, in-cluster or out-of-cluster 
```

The above example allows all traffic destined to any FQDN that matches the wildcard expression
`*foobar.com` on port 8080, originating from any Pod with label `app` set to `client`
across any Namespace. For these `client` Pods, all other egress traffic are dropped.

Note that for FQDN wildcard expressions, the `*` character can match multiple subdomains (i.e.
`*foobar.com` will match `foobar.com`, `www.foobar.com` and `test.uswest.foobar.com`).

> This feature only works at the L3/L4 level, which means that Antrea will only program datapath

Reviewer 3: I think we can remove "works at the L3/L4 level". Just say it will not apply to DNS packets.

Author: Done

      - fqdn: "svcA.default.svc.cluster.local"
```

### toServices egress rules

A combination of Service name and Service Namespace can be used in `toServices` in egress rules to refer to a K8s Service.
`toServices` match traffic based on the clusterIP, port and protocol of Services. Thus, headless Service is not supported
by this field. A sample policy can be found [here](#acnp-for-toservices-rule).

Since `toServices` represents a combination of IP+port, it cannot be used with `to` or `ports` within the same egress rule. 
Also, since the matching process relies on the groupID assigned to Service by AntreaProxy, this field can only be used when
AntreaProxy is enabled.

> This clusterIP based match has one caveat: direct access to the Endpoints of this Service is not regulated by

Reviewer 2: ```suggestion
This clusterIP-based match has one caveat: direct access to the Endpoints of this Service is not affected by
```

ipBlock will be deprecated for ipBlocks in future versions of ClusterGroup.

**ipBlocks**: This selects a list of IP CIDR ranges to allow as `ingress`
"sources" or `egress` "destinations".
A ClusterGroup with `ipBlocks` referenced in an ACNP's `appliedTo` field will be
ignored, and the policy will have no effect.
For a same ClusterGroup, `ipBlock` and `ipBlocks` cannot be set concurrently.

**serviceReference**: Pods that serve as the backend for the specified Service
will be grouped. Services without selectors are currently not supported, and will
be ignored if referred by `serviceReference` in a ClusterGroup.
When ClusterGroups with `serviceReference` are used in ACNPs as `appliedTo` or
`to`/`from` peers, no Service port information will be automatically assumed for
traffic enforcement. `ServiceReference` is merely a mechanism to group Pods and
ensure that a ClusterGroup stays in sync with the set of Pods selected by a given
Service.

**childGroups**: This selects existing ClusterGroups by name. The effective members
of the "parent" ClusterGrup will be the union of all its childGroups' members.
See the section above for restrictions.

> **status**: The ClusterGroup `status` field determines the overall realization

Reviewer 2: it is quite strange that `spec` and `status` are at the same indentation level as the other subfields, with nothing to differentiate them

Author: Added indent

It allows admins to group Pods using traditional label selectors, which can then
be referenced in ACNP in place of stand-alone `podSelector` and/or `namespaceSelector`.
In addition to `podSelector` and `namespaceSelector`, ClusterGroup also supports the
following ways to select endpoints:

- Pod grouping by `serviceReference`. ClusterGroup specified by `serviceReference` will
  contain the same Pod members that are currently selected by the Service's selector.
- `ipBlock` or `ipBlocks` to share IPBlocks between ACNPs.
- `childGroups` to select other ClusterGroups by name.

ClusterGroups allow admins to separate the concern of grouping of workloads from
the security aspect of Antrea-native policies.
It adds another level of indirection allowing users to update group membership
without having to update individual policy rules.

### The ClusterGroup resource

An example ClusterGroup might look like this:

```yaml
apiVersion: crd.antrea.io/v1alpha3
kind: ClusterGroup
metadata:
  name: test-cg-sel
spec:
  podSelector:
    matchLabels:
      role: db
  namespaceSelector:
    matchLabels:
      env: prod
status:
  conditions:
    - type: "GroupMembersComputed"
      status: "True"

>       lastTransitionTime: "2021-01-29T19:59:39Z"

Reviewer 4: I'm wondering should we keep the `status` field in an example?

Author: what do you mean?

Reviewer 4: I mean `status` is generated field. So maybe we don't need to put it in the example yaml? Like we didn't include `status` in ACNP example, etc.

Author: Make sense. Changed

The reserved label looks like: `internal.antrea.io/service-account:[ServiceAccountName]`. Users should avoid using
this label key in any entities no matter if a policy with `serviceAccount` is applied in the cluster.

## ClusterGroup

A ClusterGroup (CG) CRD is a specification of how workloads are grouped together.
It allows admins to group Pods using traditional label selectors, which can then
be referenced in ACNP in place of stand-alone `podSelector` and/or `namespaceSelector`.
In addition to `podSelector` and `namespaceSelector`, ClusterGroup also supports the
following ways to select endpoints:

- Pod grouping by `serviceReference`. ClusterGroup specified by `serviceReference` will
  contain the same Pod members that are currently selected by the Service's selector.
- `ipBlock` or `ipBlocks` to share IPBlocks between ACNPs.
- `childGroups` to select other ClusterGroups by name.

ClusterGroups allow admins to separate the concern of grouping of workloads from
the security aspect of Antrea-native policies.
It adds another level of indirection allowing users to update group membership
without having to update individual policy rules.

> ### The ClusterGroup CRD

Reviewer 3: Probably remove "The". Other section titles do not have it.

Author: Done

  priorities are only considered equal if their policy objects share the same Tier
  and have the same policy priority, plus the rules themselves are of the same rule
  priority (rule priority is the sequence number of the rule within the policy's
  ingress or egress section).
- For the "baseline" Tier, the max supported unique priorities (at rule level) is 150.
- If there are multiple Antrea-native policy rules created at the same rule-level
  priority (same policy Tier, policy priority and rule priority), who happen to select
  overlapping traffic patterns but have conflicting rule actions (e.g.`Allow`v.s.`Deny`),
  the behavior of such traffic will be undeterministic. For example, consider two
  AntreaNetworkPolicies created in the same Namespace and Tier with the same policy
  priority. The first policy applies to all `app=web` Pods in the Namespace and has only
  one ingress rule to `Deny` all traffic from `role=dev` Pods. The other policy also
  applies to all `app=web` Pods in the Namespace and has only one ingress rule, which is
  to `Allow` all traffic from `app=client` Pods. Those two ingress rules might not always

>   conflict, but in case a Pod with both `app=client` and `role=dev` labels initiate

Reviewer 2: ```suggestion
  conflict, but in case a Pod with both the `app=client` and `role=dev` labels initiates
```

>   traffic towards the web Pods in the Namespace, both rules will be matched at the same

Reviewer 2: ```suggestion
  traffic towards the `app=web` Pods in the Namespace, both rules will be matched at the same
```

  priority with conflicting actions. It will be the policy writer's responsibility to
  identify such ambiguities in rule definitions and avoid potential undeterministic
  rule enforcement results. In general, it is recommended that rules of conflicting

>   actions are always created at different rule-level priorities.

Reviewer 2: I think this is not strong enough. IMO, it should be:

> In general, it is recommended not to create rules with conflicting actions in policy resources with the same priority, if these rules may apply to the same traffic, as the behavior could be undeterministic.

And actually I think this should be the first sentence of that bullet list item (and in bold), and you can get into the details after if needed, as Jianjun suggested. The whole concept of rule level priority for rules which are defined across policies with the same priority, is an implementation detail as far as I am concerned. I don't think it is really needs to be a part of the API contract. Inserting a new rule in one of the policies could change rule enforcement order, in ways which are surprising to users.

Author: Done

  priority (same policy Tier, policy priority and rule priority), who happen to select
  overlapping traffic patterns but have conflicting rule actions (e.g.`Allow`v.s.`Deny`),
  the behavior of such traffic will be undeterministic. For example, consider two
  AntreaNetworkPolicies created in the same Namespace and Tier with the same policy
  priority. The first policy applies to all `app=web` Pods in the Namespace and has only
  one ingress rule to `Deny` all traffic from `role=dev` Pods. The other policy also
  applies to all `app=web` Pods in the Namespace and has only one ingress rule, which is
  to `Allow` all traffic from `app=client` Pods. Those two ingress rules might not always
  conflict, but in case a Pod with both `app=client` and `role=dev` labels initiate
  traffic towards the web Pods in the Namespace, both rules will be matched at the same
  priority with conflicting actions. It will be the policy writer's responsibility to
  identify such ambiguities in rule definitions and avoid potential undeterministic
  rule enforcement results. In general, it is recommended that rules of conflicting

>   actions are always created at different rule-level priorities.

Reviewer 2: honestly, I feel like we should just discourage users from defining policies at the same priority if they apply to the same Pods. Having the notion of rule-level priority across multiple policies is a bit confusing as it is not based on a user-provided priority value, but on the rule position in the list of rules.
what do you think @jianjuns @tnqn ?

Reviewer 3: We can first say the behaviors are not deterministic / predictable if two policies with the same priority apply to the same Pod, if there are both allow and deny rules. Then we can explain further the details like rule priority (assuming most readers will not read it)?

Author: I was hoping that with this paragraph it discourages people from defining policies at the same priority if they apply to the same Pods. From Antrea perspective it is hard to enforce this though.
